File: src/test/java/q1920/SolutionTest.java
package q1920;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

class SolutionTest {
    private Solution solution = new Solution();

    @Test
    void Given_singleElementArray_Should_returnSingleElementArray() {
        int[] input = {0};
        int[] actual = solution.buildArray(input);
        int[] expected = {0};
        assertArrayEquals(expected, actual);
    }

    @Test
    void Given_multipleElementArray_Should_returnArrayWithSameLength() {
        int[] input = {0, 1};
        int[] actual = solution.buildArray(input);

        assertEquals(2, actual.length);
    }

    @Test
    void test() {
        int[] input = {1, 0};
        int[] actual = solution.buildArray(input);
        int[] expected = {0, 1};
        assertArrayEquals(expected, actual);
    }

    @Test
    void testMaxLength() {
        int[] input = new int[1000];
        for (int i = 0; i < input.length; i++) {
            input[i] = i;
        }
        int[] actual = solution.buildArray(input);
        int[] expected = input;
        assertArrayEquals(expected, actual);
    }
}

File: src/test/kotlin/q1861/SolutionTest.kt
package q1861

import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory

class SolutionTest {
    private val solution = Solution()

    @TestFactory
    fun test() = listOf(
        Triple(
            "obstacle 1",
            """|*..|""",
            """
            |*|
            |.|
            |.|
            """
        ),
        Triple(
            "obstacle 2",
            """
                |*..|
                |.*.|
               """,
            """
                |.*|
                |*.|
                |..|
                """
        ),
        Triple(
            "obstacle 3",
            """
                |*..*|
                |.**.|
                """,
            """
                |.*|
                |*.|
                |*.|
                |.*|
                """
        ),
        Triple(
            "stone 1",
            "|#..|",
            """
            |.|
            |.|
            |#|
            """
        ),
        Triple(
            "stone 2",
            """
                |###|
                |##.|
               """,
            """
                |.#|
                |##|
                |##|
                """
        ),
        Triple(
            "stone 3",
            """
                |#..#|
                |.##.|
                """,
            """
                |..|
                |..|
                |##|
                |##|
                """
        ),
        Triple(
            "stone and obstacle 1",
            "|#*.|",
            """
            |#|
            |*|
            |.|
            """
        ),
        Triple(
            "stone and obstacle 2",
            """
                |#*#.|
                |#.*#|
               """,
            """
                |.#|
                |#*|
                |*.|
                |##|
                """
        ),
        Triple(
            "stone and obstacle 3",
            """
                |#.#..***|
                |#.*#.*#.|
                """,
            """
                |..|
                |#.|
                |*.|
                |.#|
                |##|
                |**|
                |.*|
                |#*|
                """
        ),
    ).map { (displayName, input, expected) ->
        DynamicTest.dynamicTest(displayName) {
            assertArrayEquals(expected.toBox(), solution.rotateTheBox(input.toBox()))
        }
    }

    private fun String.toBox(): Array<CharArray> {
        return this.trimIndent().replace("|", "").split("\n").map { s ->
            s.toCharArray()
        }.toTypedArray()
    }
}